File: packages/core/src/gen-ai-attributes.ts
/**
 * OTel GenAI Semantic Conventions — attribute helpers.
 *
 * The OTel spec working group standardized `gen_ai.*` attributes for LLM /
 * agent telemetry so any OTel backend (Datadog, Honeycomb, Jaeger, Smoo's
 * own LLM dashboard) can interpret them with one shared vocabulary. This
 * module exports a typed helper for setting those attributes on spans
 * without hand-coding the constants in every caller.
 *
 * Spec reference: https://opentelemetry.io/docs/specs/semconv/gen-ai/
 *
 * SMOODEV-1155 — initial vocabulary. The spec is still moving (recent
 * additions cover tool calls + streaming + multi-agent loops); extend
 * `GenAIAttributes` as new attributes stabilize.
 */
import type { Span } from '@opentelemetry/api';

export type GenAIOperationName = 'chat' | 'text_completion' | 'embeddings' | 'tool' | 'agent' | 'rerank';

export type GenAISystem = 'openai' | 'anthropic' | 'groq' | 'gemini' | 'cohere' | 'mistral' | 'deepseek' | 'azure_openai' | 'aws_bedrock' | (string & {});

/**
 * Full set of GenAI semantic-convention attributes. All optional — set what
 * you know. Used by `setGenAIAttributes(span, attrs)` to apply onto an
 * active OTel span. Keys mirror the spec exactly so backends parse without
 * mapping.
 */
export interface GenAIAttributes {
    /** Provider system, e.g. 'openai', 'anthropic'. */
    system?: GenAISystem;
    /** Operation kind. Pick 'agent' for top-level agent spans, 'chat' for LLM calls, 'tool' for tool invocations. */
    operationName?: GenAIOperationName;

    /** Model the request asked for, e.g. 'gpt-4o', 'claude-opus-4-7'. */
    requestModel?: string;
    /** Model the response actually came back from (may differ from requested under fallback). */
    responseModel?: string;
    /** Provider-issued ID of the response (Anthropic message ID, OpenAI completion ID, etc.). */
    responseId?: string;

    /** Sampling temperature. */
    temperature?: number;
    /** Top-P sampling cutoff. */
    topP?: number;
    /** Top-K sampling cutoff. */
    topK?: number;
    /** Max generated token budget. */
    maxTokens?: number;
    /** Random seed if supplied. */
    seed?: number;

    /** Input/prompt tokens billed. */
    usageInputTokens?: number;
    /** Output/completion tokens billed. */
    usageOutputTokens?: number;
    /** Cached-prompt tokens (Anthropic prompt-cache hit / OpenAI cached input). */
    usageCachedTokens?: number;
    /** Total cost in USD (filled in by the platform from upstream provider invoices when available). */
    usageCostUsd?: number;

    /** Comma-separated list of tool names called within this span. */
    toolNames?: string[];
    /** Truncation flag — true if the response was cut off by max_tokens / stop_reason. */
    truncated?: boolean;
    /** Provider's reported finish reason ('stop', 'length', 'tool_calls', etc.). */
    finishReason?: string;

    /** End-user id from your system (NOT the provider's id). */
    endUserId?: string;
    /** Conversation/session id. */
    conversationId?: string;
}

/**
 * Apply a {@link GenAIAttributes} value onto the active OTel span. Skips
 * undefined fields so partial calls (e.g. setting just the model + tokens
 * after a streaming completion finishes) are idempotent.
 */
export function setGenAIAttributes(span: Span, attrs: GenAIAttributes): void {
    if (attrs.system !== undefined) span.setAttribute('gen_ai.system', attrs.system);
    if (attrs.operationName !== undefined) span.setAttribute('gen_ai.operation.name', attrs.operationName);
    if (attrs.requestModel !== undefined) span.setAttribute('gen_ai.request.model', attrs.requestModel);
    if (attrs.responseModel !== undefined) span.setAttribute('gen_ai.response.model', attrs.responseModel);
    if (attrs.responseId !== undefined) span.setAttribute('gen_ai.response.id', attrs.responseId);
    if (attrs.temperature !== undefined) span.setAttribute('gen_ai.request.temperature', attrs.temperature);
    if (attrs.topP !== undefined) span.setAttribute('gen_ai.request.top_p', attrs.topP);
    if (attrs.topK !== undefined) span.setAttribute('gen_ai.request.top_k', attrs.topK);
    if (attrs.maxTokens !== undefined) span.setAttribute('gen_ai.request.max_tokens', attrs.maxTokens);
    if (attrs.seed !== undefined) span.setAttribute('gen_ai.request.seed', attrs.seed);
    if (attrs.usageInputTokens !== undefined) span.setAttribute('gen_ai.usage.input_tokens', attrs.usageInputTokens);
    if (attrs.usageOutputTokens !== undefined) span.setAttribute('gen_ai.usage.output_tokens', attrs.usageOutputTokens);
    if (attrs.usageCachedTokens !== undefined) span.setAttribute('gen_ai.usage.cached_tokens', attrs.usageCachedTokens);
    if (attrs.usageCostUsd !== undefined) span.setAttribute('gen_ai.usage.cost_usd', attrs.usageCostUsd);
    if (attrs.toolNames !== undefined && attrs.toolNames.length > 0) span.setAttribute('gen_ai.tool.names', attrs.toolNames);
    if (attrs.truncated !== undefined) span.setAttribute('gen_ai.response.truncated', attrs.truncated);
    if (attrs.finishReason !== undefined) span.setAttribute('gen_ai.response.finish_reason', attrs.finishReason);
    if (attrs.endUserId !== undefined) span.setAttribute('gen_ai.end_user.id', attrs.endUserId);
    if (attrs.conversationId !== undefined) span.setAttribute('gen_ai.conversation.id', attrs.conversationId);
}

/**
 * Emit a `gen_ai.user.message` / `gen_ai.assistant.message` / `gen_ai.system.message`
 * span event so the dashboard's prompt/completion side-by-side view can render
 * the actual content of the LLM call. Use sparingly — these are size-heavy.
 */
export function recordGenAIMessage(
    span: Span,
    role: 'user' | 'assistant' | 'system' | 'tool',
    content: string,
    extra?: { toolCallId?: string; toolName?: string },
): void {
    const eventName = `gen_ai.${role}.message`;
    span.addEvent(eventName, {
        'gen_ai.message.content': content,
        ...(extra?.toolCallId !== undefined && { 'gen_ai.tool_call.id': extra.toolCallId }),
        ...(extra?.toolName !== undefined && { 'gen_ai.tool.name': extra.toolName }),
    });
}

File: dotnet/src/SmooAI.Observability/GenAI/GenAIAttributes.cs
using System.Diagnostics;

namespace SmooAI.Observability.GenAI;

/// <summary>
/// OTel GenAI semantic-convention attributes. All optional — set what you know.
/// Keys mirror the spec exactly (<c>gen_ai.*</c>) so any GenAI-semconv-aware OTel
/// backend (Datadog, Honeycomb, Phoenix, Smoo's LLM dashboard) parses them
/// without mapping. Port of the TS <c>gen-ai-attributes.ts</c>.
///
/// Spec: https://opentelemetry.io/docs/specs/semconv/gen-ai/
/// </summary>
public sealed class GenAIAttributes
{
    /// <summary>Provider system, e.g. 'openai', 'anthropic'.</summary>
    public string? System { get; set; }

    /// <summary>Operation kind — 'agent', 'chat', 'tool', 'embeddings', etc.</summary>
    public string? OperationName { get; set; }

    /// <summary>Model the request asked for, e.g. 'gpt-4o', 'claude-opus-4-7'.</summary>
    public string? RequestModel { get; set; }

    /// <summary>Model the response actually came from (may differ under fallback).</summary>
    public string? ResponseModel { get; set; }

    /// <summary>Provider-issued response id.</summary>
    public string? ResponseId { get; set; }

    /// <summary>Sampling temperature.</summary>
    public double? Temperature { get; set; }

    /// <summary>Top-P sampling cutoff.</summary>
    public double? TopP { get; set; }

    /// <summary>Top-K sampling cutoff.</summary>
    public double? TopK { get; set; }

    /// <summary>Max generated token budget.</summary>
    public int? MaxTokens { get; set; }

    /// <summary>Random seed, if supplied.</summary>
    public long? Seed { get; set; }

    /// <summary>Input/prompt tokens billed.</summary>
    public int? UsageInputTokens { get; set; }

    /// <summary>Output/completion tokens billed.</summary>
    public int? UsageOutputTokens { get; set; }

    /// <summary>Cached-prompt tokens (prompt-cache hit / cached input).</summary>
    public int? UsageCachedTokens { get; set; }

    /// <summary>Total cost in USD, when available.</summary>
    public double? UsageCostUsd { get; set; }

    /// <summary>Tool names called within this span.</summary>
    public IReadOnlyList<string>? ToolNames { get; set; }

    /// <summary>Truncation flag — true if the response was cut off.</summary>
    public bool? Truncated { get; set; }

    /// <summary>Provider's reported finish reason ('stop', 'length', 'tool_calls', ...).</summary>
    public string? FinishReason { get; set; }

    /// <summary>End-user id from your system (not the provider's).</summary>
    public string? EndUserId { get; set; }

    /// <summary>Conversation/session id.</summary>
    public string? ConversationId { get; set; }
}

/// <summary>
/// Helpers to apply <see cref="GenAIAttributes"/> onto an <see cref="Activity"/>
/// (.NET's span type). Skips unset fields so partial calls are idempotent. Port
/// of the TS <c>setGenAIAttributes</c> / <c>recordGenAIMessage</c>.
/// </summary>
public static class GenAIActivity
{
    /// <summary>Apply GenAI attributes onto the given span. No-op if span is null.</summary>
    public static void SetAttributes(Activity? span, GenAIAttributes attrs)
    {
        if (span is null || attrs is null)
        {
            return;
        }
        SetIf(span, "gen_ai.system", attrs.System);
        SetIf(span, "gen_ai.operation.name", attrs.OperationName);
        SetIf(span, "gen_ai.request.model", attrs.RequestModel);
        SetIf(span, "gen_ai.response.model", attrs.ResponseModel);
        SetIf(span, "gen_ai.response.id", attrs.ResponseId);
        SetIf(span, "gen_ai.request.temperature", attrs.Temperature);
        SetIf(span, "gen_ai.request.top_p", attrs.TopP);
        SetIf(span, "gen_ai.request.top_k", attrs.TopK);
        SetIf(span, "gen_ai.request.max_tokens", attrs.MaxTokens);
        SetIf(span, "gen_ai.request.seed", attrs.Seed);
        SetIf(span, "gen_ai.usage.input_tokens", attrs.UsageInputTokens);
        SetIf(span, "gen_ai.usage.output_tokens", attrs.UsageOutputTokens);
        SetIf(span, "gen_ai.usage.cached_tokens", attrs.UsageCachedTokens);
        SetIf(span, "gen_ai.usage.cost_usd", attrs.UsageCostUsd);
        if (attrs.ToolNames is { Count: > 0 })
        {
            span.SetTag("gen_ai.tool.names", attrs.ToolNames.ToArray());
        }
        SetIf(span, "gen_ai.response.truncated", attrs.Truncated);
        SetIf(span, "gen_ai.response.finish_reason", attrs.FinishReason);
        SetIf(span, "gen_ai.end_user.id", attrs.EndUserId);
        SetIf(span, "gen_ai.conversation.id", attrs.ConversationId);
    }

    /// <summary>
    /// Emit a <c>gen_ai.{role}.message</c> span event carrying the message content
    /// so the dashboard's prompt/completion view can render it. Use sparingly —
    /// these are size-heavy. No-op if span is null.
    /// </summary>
    public static void RecordMessage(Activity? span, string role, string content, string? toolCallId = null, string? toolName = null)
    {
        if (span is null)
        {
            return;
        }
        var tags = new ActivityTagsCollection
        {
            { "gen_ai.message.content", content },
        };
        if (toolCallId is not null)
        {
            tags["gen_ai.tool_call.id"] = toolCallId;
        }
        if (toolName is not null)
        {
            tags["gen_ai.tool.name"] = toolName;
        }
        span.AddEvent(new ActivityEvent($"gen_ai.{role}.message", tags: tags));
    }

    private static void SetIf(Activity span, string key, string? value)
    {
        if (value is not null)
        {
            span.SetTag(key, value);
        }
    }

    private static void SetIf(Activity span, string key, double? value)
    {
        if (value.HasValue)
        {
            span.SetTag(key, value.Value);
        }
    }

    private static void SetIf(Activity span, string key, int? value)
    {
        if (value.HasValue)
        {
            span.SetTag(key, value.Value);
        }
    }

    private static void SetIf(Activity span, string key, long? value)
    {
        if (value.HasValue)
        {
            span.SetTag(key, value.Value);
        }
    }

    private static void SetIf(Activity span, string key, bool? value)
    {
        if (value.HasValue)
        {
            span.SetTag(key, value.Value);
        }
    }
}
